test/e2e/chaos_test.go

```diff
@@ -0,0 +1,205 @@
+// +build e2e
+
+package e2e
+
+import (
+	"context"
+	"math/rand"
+	"sync"
+	"testing"
+	"time"
+
+	. "github.com/onsi/gomega"
+	cpomanifests "github.com/openshift/hypershift/control-plane-operator/controllers/hostedcontrolplane/manifests"
+	"github.com/openshift/hypershift/hypershift-operator/controllers/manifests"
+	"k8s.io/apimachinery/pkg/labels"
+	"k8s.io/apimachinery/pkg/util/wait"
+	"k8s.io/utils/pointer"
+
+	hyperv1 "github.com/openshift/hypershift/api/v1alpha1"
+	e2eutil "github.com/openshift/hypershift/test/e2e/util"
+	corev1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	crclient "sigs.k8s.io/controller-runtime/pkg/client"
+)
+
+// TestEtcdChaos launches an HA control plane and executes a suite of chaotic
+// etcd tests which ensure no data is lost in the chaos.
+func TestEtcdChaos(t *testing.T) {
+	ctx, cancel := context.WithCancel(testContext)
+	defer cancel()
+
+	client := e2eutil.GetClientOrDie()
+
+	// Create a cluster
+	clusterOpts := globalOpts.DefaultClusterOptions()
+	clusterOpts.ControlPlaneAvailabilityPolicy = string(hyperv1.HighlyAvailable)
+	clusterOpts.NodePoolReplicas = 0
+
+	cluster, teardownCluster := e2eutil.CreateCluster(t, ctx, client, clusterOpts, globalOpts.ArtifactDir)
+	defer teardownCluster(context.Background())
+
+	t.Run("KillRandomMembers", testKillRandomMembers(ctx, client, cluster))
+	t.Run("KillAllMembers", testKillAllMembers(ctx, client, cluster))
+}
+
+// testKillRandomMembers ensures that data is preserved following a period where
+// random etcd members are repeatedly killed.
+func testKillRandomMembers(parentCtx context.Context, client crclient.Client, cluster *hyperv1.HostedCluster) func(t *testing.T) {
+	return func(t *testing.T) {
+		g := NewWithT(t)
+		ctx, cancel := context.WithCancel(parentCtx)
+		defer cancel()
+
+		guestNamespace := manifests.HostedControlPlaneNamespace(cluster.Namespace, cluster.Name).Name
+		t.Logf("Hosted control plane namespace is %s", guestNamespace)
+
+		// Get a client for the cluster
+		t.Logf("Waiting for guest client to become available")
+		guestClient := e2eutil.WaitForGuestClient(t, testContext, client, cluster)
+
+		// Create data in the cluster which should survive the ensuring chaos
+		value, _ := time.Now().MarshalText()
+		cm := &corev1.ConfigMap{
```

Comment:
This isn't testing etcd directly in that way, it's testing our operational implementation. It checks that data isn't lost because of how we've deployed and are operating etcd. e.g. does the persistent volume configuration actually work, are we assuring data isn't lost due to some behavioral error in the orchestration code, etc.